In [1]:
# To enable plotting graphs in Jupyter notebook
%matplotlib inline

In [108]:
import pandas as pd
from sklearn.linear_model import LogisticRegression

# importing ploting libraries
import matplotlib.pyplot as plt   

#importing seaborn for statistical plots
import seaborn as sns

#Let us break the X and y dataframes into training set and test set. For this we will use
#Sklearn package's data splitting function which is based on random function

from sklearn.model_selection import train_test_split

import numpy as np
#import os,sys
from scipy import stats

# calculate accuracy measures and confusion matrix
import pandas
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn import preprocessing

In [109]:
df = pd.read_csv('banking_updated.csv')
df.head()

age          job  marital          education  default housing loan  \
0   44  blue-collar  married           basic.4y  unknown     yes   no   
1   53   technician  married            unknown       no      no   no   
2   28   management   single  university.degree       no     yes   no   
3   39     services  married        high.school       no      no   no   
4   55      retired  married           basic.4y       no     yes   no   

    contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0  cellular   aug         thu  ...         1    999         0  nonexistent   
1  cellular   nov         fri  ...         1    999         0  nonexistent   
2  cellular   jun         thu  ...         3      6         2      success   
3  cellular   apr         fri  ...         2    999         0  nonexistent   
4  cellular   aug         fri  ...         1      3         1      success   

  emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed  y  
0          1.4          93.444          -36.1      4.963       5228.1  0  
1         -0.1          93.200          -42.0      4.021       5195.8  0  
2         -1.7          94.055          -39.8      0.729       4991.6  1  
3         -1.8          93.075          -47.1      1.405       5099.1  0  
4         -2.9          92.201          -31.4      0.869       5076.2  1  

[5 rows x 21 columns]

In [110]:
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp_var_rate  cons_price_idx  cons_conf_idx     euribor3m  \
count  41188.000000    41188.000000   41188.000000  41188.000000   
mean       0.081886       93.575664     -40.502600      3.621291   
std        1.570960        0.578840       4.628198      1.734447   
min       -3.400000       92.201000     -50.800000      0.634000   
25%       -1.800000       93.075000     -42.700000      1.344000   
50%        1.100000       93.749000     -41.800000      4.857000   
75%        1.400000       93.994000     -36.400000      4.961000   
max        1.400000       94.767000     -26.900000      5.045000   

        nr_employed             y  
count  41188.000000  41188.000000  
mean    5167.035911      0.112654  
std       72.251528      0.316173  
min     4963.600000      0.000000  
25%     5099.100000      0.000000  
50%     5191.000000      0.000000  
75%     5228.100000      0.000000  
max     5228.100000      1.000000

**Input variables:**

1 - age (numeric)  
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')  
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)  
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')  
5 - default: has credit in default? (categorical: 'no','yes','unknown')  
6 - housing: has housing loan? (categorical: 'no','yes','unknown')  
7 - loan: has personal loan? (categorical: 'no','yes','unknown')  
8 - contact: contact communication type (categorical: 'cellular','telephone')  
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')  
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')  
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no')  
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)  
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)  
14 - previous: number of contacts performed before this campaign and for this client (numeric)  
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')  
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)  
17 - cons.price.idx: consumer price index - monthly indicator (numeric)  
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)  
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)  
20 - nr.employed: number of employees - quarterly indicator (numeric)  

**Output variable (desired target):**  
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')

In [111]:
df.drop(['duration','contact','month','day_of_week','default','pdays',],axis=1,inplace=True)

In [112]:
df.head()

age          job  marital          education housing loan  campaign  \
0   44  blue-collar  married           basic.4y     yes   no         1   
1   53   technician  married            unknown      no   no         1   
2   28   management   single  university.degree     yes   no         3   
3   39     services  married        high.school      no   no         2   
4   55      retired  married           basic.4y     yes   no         1   

   previous     poutcome  emp_var_rate  cons_price_idx  cons_conf_idx  \
0         0  nonexistent           1.4          93.444          -36.1   
1         0  nonexistent          -0.1          93.200          -42.0   
2         2      success          -1.7          94.055          -39.8   
3         0  nonexistent          -1.8          93.075          -47.1   
4         1      success          -2.9          92.201          -31.4   

   euribor3m  nr_employed  y  
0      4.963       5228.1  0  
1      4.021       5195.8  0  
2      0.729       4991.6  1  
3      1.405       5099.1  0  
4      0.869       5076.2  1

In [113]:
df.replace(['basic.6y','basic.4y', 'basic.9y'], 'basic', inplace=True)

### Pre Processing
Sklearn provides a very efficient tool for encoding the levels of a categorical features into numeric values. `LabelEncoder` encode labels with value between 0 and n_classes-1

In [114]:
le = preprocessing.LabelEncoder()

In [115]:
df.job = le.fit_transform(df.job)

In [116]:
df.marital = le.fit_transform(df.marital)

In [117]:
df.education = le.fit_transform(df.education)
df.housing = le.fit_transform(df.housing)
df.loan = le.fit_transform(df.loan)
df.poutcome = le.fit_transform(df.poutcome)

In [118]:
df.head()

age  job  marital  education  housing  loan  campaign  previous  poutcome  \
0   44    1        1          0        2     0         1         0         1   
1   53    9        1          5        0     0         1         0         1   
2   28    4        2          4        2     0         3         2         2   
3   39    7        1          1        0     0         2         0         1   
4   55    5        1          0        2     0         1         1         2   

   emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed  y  
0           1.4          93.444          -36.1      4.963       5228.1  0  
1          -0.1          93.200          -42.0      4.021       5195.8  0  
2          -1.7          94.055          -39.8      0.729       4991.6  1  
3          -1.8          93.075          -47.1      1.405       5099.1  0  
4          -2.9          92.201          -31.4      0.869       5076.2  1

In [119]:
df.shape

(41188, 15)

In [120]:
X = df.iloc[:,0:14]
X[0:10]

age  job  marital  education  housing  loan  campaign  previous  poutcome  \
0   44    1        1          0        2     0         1         0         1   
1   53    9        1          5        0     0         1         0         1   
2   28    4        2          4        2     0         3         2         2   
3   39    7        1          1        0     0         2         0         1   
4   55    5        1          0        2     0         1         1         2   
5   30    4        0          0        2     0         8         0         1   
6   37    1        1          0        2     0         1         0         1   
7   39    1        0          0        2     0         1         0         1   
8   36    0        1          4        0     0         1         1         2   
9   27    1        2          0        2     0         2         1         0   

   emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed  
0           1.4          93.444          -36.1      4.963       5228.1  
1          -0.1          93.200          -42.0      4.021       5195.8  
2          -1.7          94.055          -39.8      0.729       4991.6  
3          -1.8          93.075          -47.1      1.405       5099.1  
4          -2.9          92.201          -31.4      0.869       5076.2  
5           1.4          93.918          -42.7      4.961       5228.1  
6          -1.8          92.893          -46.2      1.327       5099.1  
7          -1.8          92.893          -46.2      1.313       5099.1  
8          -2.9          92.963          -40.8      1.266       5076.2  
9          -1.8          93.075          -47.1      1.410       5099.1

In [121]:
Y = df.iloc[:,14]
Y[0:10]

0    0
1    0
2    1
3    0
4    1
5    0
6    0
7    0
8    1
9    0
Name: y, dtype: int64

In [122]:
#sns.pairplot(df,diag_kind='kde')

In [123]:
df.replace(['basic.6y','basic.4y', 'basic.9y'], 'basic', inplace=True)

In [124]:
df.head()

age  job  marital  education  housing  loan  campaign  previous  poutcome  \
0   44    1        1          0        2     0         1         0         1   
1   53    9        1          5        0     0         1         0         1   
2   28    4        2          4        2     0         3         2         2   
3   39    7        1          1        0     0         2         0         1   
4   55    5        1          0        2     0         1         1         2   

   emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed  y  
0           1.4          93.444          -36.1      4.963       5228.1  0  
1          -0.1          93.200          -42.0      4.021       5195.8  0  
2          -1.7          94.055          -39.8      0.729       4991.6  1  
3          -1.8          93.075          -47.1      1.405       5099.1  0  
4          -2.9          92.201          -31.4      0.869       5076.2  1

### Train and Test split
A general practice is to split your data into a training and test set. You train/tune your model with your training set and test how well it generalizes to data it has never seen before with your test set. 
We will make use of scikit learn’s `train_test_split` method to achieve this

In [125]:
array=df.values
X = df.iloc[:,0:14]
y = df.iloc[:,14]
test_size=0.30 #taking 70:30 training and test set
seed =7 #random number seeding for repeatability of the code.
x_train, x_test, y_train, y_test = model_selection.train_test_split(X, y, test_size=test_size, random_state=0)


In [126]:
y[0:10]

0    0
1    0
2    1
3    0
4    1
5    0
6    0
7    0
8    1
9    0
Name: y, dtype: int64

In [127]:
X[0:10]

age  job  marital  education  housing  loan  campaign  previous  poutcome  \
0   44    1        1          0        2     0         1         0         1   
1   53    9        1          5        0     0         1         0         1   
2   28    4        2          4        2     0         3         2         2   
3   39    7        1          1        0     0         2         0         1   
4   55    5        1          0        2     0         1         1         2   
5   30    4        0          0        2     0         8         0         1   
6   37    1        1          0        2     0         1         0         1   
7   39    1        0          0        2     0         1         0         1   
8   36    0        1          4        0     0         1         1         2   
9   27    1        2          0        2     0         2         1         0   

   emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed  
0           1.4          93.444          -36.1      4.963       5228.1  
1          -0.1          93.200          -42.0      4.021       5195.8  
2          -1.7          94.055          -39.8      0.729       4991.6  
3          -1.8          93.075          -47.1      1.405       5099.1  
4          -2.9          92.201          -31.4      0.869       5076.2  
5           1.4          93.918          -42.7      4.961       5228.1  
6          -1.8          92.893          -46.2      1.327       5099.1  
7          -1.8          92.893          -46.2      1.313       5099.1  
8          -2.9          92.963          -40.8      1.266       5076.2  
9          -1.8          93.075          -47.1      1.410       5099.1

In [128]:
x_train.shape, y_train.shape

((28831, 14), (28831,))

In [129]:
x_test.shape, y_test.shape

((12357, 14), (12357,))

### Training the model
We will make use of different classification algorithms to train this data set and will record the accuracy on test set.

#### Logistic Regression

In [130]:
model=LogisticRegression(penalty='l2', max_iter=1000)

In [131]:
model.fit(x_train, y_train)

C:\Users\arun.gopalakrishnan\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1000,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [132]:
prediction = model.predict(X_test)

In [133]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_test, prediction)

0.9013514607105284

In [134]:
prediction

array([0, 0, 0, ..., 0, 0, 1], dtype=int64)

In [107]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(Y_test, prediction)
print(confusion_matrix)

[[10903    80]
 [ 1146   228]]
